src/tfs/tlog.c

```diff
@@ -1,180 +1,418 @@
-                    tlog_debug("   found fsfile %p\n", f);
-                }
-            } else if (k == sym(filelength)) {
-                assert(u64_from_value(v, &filelength));
+            /* chain to next log extension, carrying status handler to end */
+            read_log(tl, sh);
+            goto out;
+        case TUPLE_AVAILABLE:
+            tlog_debug("-> tuple available\n");
+            if (tl->tuple_bytes_remain > 0) {
+                s = timm("result", "TUPLE_AVAILABLE read while already parsing tuple (%ld remaining)",
+                         tl->tuple_bytes_remain);
+                goto out_apply_status;
             }
-        }
-        
-        if (f && filelength != infinity) {
-            tlog_debug("   update fsfile length to %ld\n", filelength);
-            fsfile_set_length(f, filelength);
+            tuple_length = pop_varint(b);
+            length = pop_varint(b); /* of segment */
+            tlog_debug("tuple total length %ld, available %ld\n", tuple_length, length);
+            if (length > tuple_length || length + 1 > buffer_length(b)) {
+                s = timm("result", "TUPLE_AVAILABLE read with invalid or short available "
+                         "length (%ld, tuple_length: %ld, buffer length: %ld)",
+                         length, tuple_length, buffer_length(b));
+                goto out_apply_status;
+            }
+            if (length == tuple_length) {
+                /* read at once from log staging */
+                log_parse_tuple(tl, b);
+            } else {
+                /* this tuple is in installments */
+                buffer_clear(tl->tuple_staging);
+                tl->tuple_bytes_remain = tuple_length;
+                log_tuple_produce(tl, b, length);
+            }
+            break;
+        case TUPLE_EXTENDED:
+            tlog_debug("-> tuple extended data\n");
+            length = pop_varint(b);
+            if (length > tl->tuple_bytes_remain) {
+                s = timm("result", "TUPLE_EXTENDED read of length %ld, exceeding remainder %ld",
+                         length, tl->tuple_bytes_remain);
+                goto out_apply_status;
+            }
+            tlog_debug("need %ld, available %ld\n", tl->tuple_bytes_remain, length);
+            log_tuple_produce(tl, b, length);
+            if (tl->tuple_bytes_remain == 0) {
+                log_parse_tuple(tl, tl->tuple_staging);
+                buffer_clear(tl->tuple_staging);
+            }
+            break;
+        default:
+            tlog_debug("-> unknown encoding type %d\n", frame);
+            s = timm("result", "unknown frame identifier 0x%x", frame);
+            goto out_apply_status;
         }
     }
 
-    if (frame == END_OF_LOG) {
-        *(u8*)(b->contents + b->start - 1) = END_OF_SEGMENT;
-    }
+    assert(frame == END_OF_LOG);
+    tlog_debug("-> end of log\n");
+
+    /* the log must go on */
+    *(u8*)(b->contents + b->start - 1) = END_OF_SEGMENT;
```

Comment:
I know this is pre-existing...but doesn't it seem like we should kick of the laziest and most coalescing flush possible here?

Reply:
For sure...as it is now, an explicit flush is happening immediately after each filesystem operation.